src/Microsoft.Diagnostics.NETCore.Client/DiagnosticsIpc/IpcTransport.cs

```diff
@@ -0,0 +1,418 @@
+        Task WaitForConnectionAsync(CancellationToken token);
+
+        /// <summary>
+        /// Connects to the underlying IPC Transport and opens a read/write-able Stream
+        /// </summary>
+        /// <returns>A Stream for writing and reading data to and from the target .NET process</returns>
+        Stream Connect();
+    }
+
+    internal abstract class BaseIpcEndpoint : IIpcEndpoint
+    {
+        private readonly Queue<StreamTarget> _targets = new Queue<StreamTarget>();
+
+        private bool _disposed;
+        private Stream _stream;
+
+        /// <inheritdoc cref="IIpcEndpoint.Connect"/>
+        /// <remarks>
+        /// This will block until the diagnostic stream is provided. This block can happen if
+        /// the stream is acquired previously and the runtime instance has not yet reconnected
+        /// to the reversed diagnostics server.
+        /// </remarks>
+        public Stream Connect()
+        {
+            using var target = new StreamEventTarget();
+
+            RegisterTarget(target);
+
+            target.Handle.WaitOne();
+
+            return target.Stream;
+        }
+
+        /// <inheritdoc cref="IIpcEndpoint.WaitForConnectionAsync(CancellationToken)"/>
+        public async Task WaitForConnectionAsync(CancellationToken token)
+        {
+            bool isConnected = false;
+            do
+            {
+                using (var target = new StreamTaskTarget(token))
+                {
+                    Stream stream = null;
+                    try
+                    {
+                        RegisterTarget(target);
+
+                        // Wait for a stream to be provided by the target
+                        stream = await target.Task;
+
+                        // Test if the stream is viable
+                        isConnected = TestStream(stream);
+                    }
+                    catch (Exception ex) when (!(ex is OperationCanceledException))
+                    {
+                        // Handle all exceptions except cancellation
+                    }
+                    finally
+                    {
+                        // If the stream is not connected, it can be disposed;
+                        // otherwise, make the stream available again since waiting
+                        // for the stream to connect should not consume the stream.
+                        if (!isConnected)
+                        {
+                            stream?.Dispose();
+                        }
+                        else
+                        {
+                            ProvideStream(stream);
+                        }
+                    }
+                }
+
+                // If the stream is not connected, wait briefly to allow
+                // the runtime instance to possibly repopulate a new connection stream.
+                if (!isConnected)
+                {
+                    await Task.Delay(TimeSpan.FromMilliseconds(20), token);
```

Comment:
nit: consider moving all timeouts to a more central place that can easily be tweaked.